persistence/sql/process_wait_until.go

```diff
@@ -0,0 +1,184 @@
+		return nil, err
+	}
+
+	if len(request.PublishToLocalQueue) > 0 {
+		hasNewImmediateTask = true
+	}
+
+	return &persistence.ProcessWaitUntilExecutionResponse{
+		HasNewImmediateTask: hasNewImmediateTask,
+	}, nil
+}
+
+func (p sqlProcessStoreImpl) CompleteWaitUntilExecution(
+	ctx context.Context, tx extensions.SQLTransaction, request persistence.CompleteWaitUntilExecutionRequest,
+) error {
+	stateRow := extensions.AsyncStateExecutionRowForUpdateWithoutCommands{
+		ProcessExecutionId: request.ProcessExecutionId,
+		StateId:            request.StateId,
+		StateIdSequence:    request.StateIdSequence,
+		WaitUntilStatus:    persistence.StateExecutionStatusCompleted,
+		ExecuteStatus:      persistence.StateExecutionStatusRunning,
+		PreviousVersion:    request.PreviousVersion,
+		LastFailure:        nil,
+	}
+
+	err := tx.UpdateAsyncStateExecutionWithoutCommands(ctx, stateRow)
+	if err != nil {
+		if p.session.IsConditionalUpdateFailure(err) {
+			p.logger.Warn("UpdateAsyncStateExecutionWithoutCommands failed at conditional update")
+		}
+		return err
+	}
+
+	return tx.InsertImmediateTask(ctx, extensions.ImmediateTaskRowForInsert{
+		ShardId:            request.TaskShardId,
+		TaskType:           persistence.ImmediateTaskTypeExecute,
+		ProcessExecutionId: request.ProcessExecutionId,
+		StateId:            request.StateId,
+		StateIdSequence:    request.StateIdSequence,
+	})
+}
+
+func (p sqlProcessStoreImpl) updateWaitUntilExecution(
+	ctx context.Context, tx extensions.SQLTransaction, request persistence.ProcessWaitUntilExecutionRequest,
+) error {
+	commandRequestBytes, err := persistence.FromCommandRequestToBytes(request.CommandRequest)
+	if err != nil {
+		return err
+	}
+
+	commandResultsBytes, err := persistence.FromCommandResultsToBytes(xdbapi.CommandResults{})
+	if err != nil {
+		return err
+	}
+
+	stateRow := extensions.AsyncStateExecutionRowForUpdate{
+		ProcessExecutionId: request.ProcessExecutionId,
+
+		StateId:         request.StateId,
+		StateIdSequence: request.StateIdSequence,
+
+		WaitUntilStatus: persistence.StateExecutionStatusWaitingForCommands,
+		ExecuteStatus:   persistence.StateExecutionStatusUndefined,
+
+		WaitUntilCommands:       commandRequestBytes,
+		WaitUntilCommandResults: commandResultsBytes,
+
+		LastFailure: nil,
+
+		PreviousVersion: request.Prepare.PreviousVersion,
+	}
+
+	localQueueCommands := request.CommandRequest.GetLocalQueueCommands()
+
+	// lock and handle process execution row first
+	if len(localQueueCommands) > 0 {
+		prcRow, err := tx.SelectProcessExecutionForUpdate(ctx, request.ProcessExecutionId)
+		if err != nil {
+			return err
+		}
+
+		waitingQueues, err := persistence.NewStateExecutionWaitingQueuesFromBytes(prcRow.StateExecutionWaitingQueues)
+		if err != nil {
+			return err
+		}
+
+		for _, localQueueCommand := range localQueueCommands {
+			waitingQueues.Add(request.StateExecutionId, localQueueCommand, request.CommandRequest.GetWaitingType() == xdbapi.ANY_OF_COMPLETION)
```

Comment:
I think it make more sense to add for the whole command request -- the waitUntil API only returns once for those commands. it only have one waitingType(AnyOf or AllOf today)

Reply:
Here, the whole local queue command requests have been added into `waitingQueues`. I am not going to support timer request in thie PR.